Bluepill-cli/bp_test_dumper/bp_test_dumper.m

```diff
@@ -0,0 +1,90 @@
+    NSLog(@"[BPTestDumper] Got a total of %lu xctest bundles.", [xcTests count]);
+    for (NSString *xcTestBundlePath in xcTests) {
+        NSString *ext = [xcTestBundlePath pathExtension];
+        if (![ext isEqual:@"xctest"]) {
+            NSLog(@"[BPTestDumper] Skipping %@ - %@", xcTestBundlePath, ext);
+            continue;
+        }
+        NSString *path = [pluginsPath stringByAppendingPathComponent:xcTestBundlePath];
+        NSLog(@"[BPTestDumper] Loading %@", path);
+        NSBundle *bundle = [NSBundle bundleWithPath:path];
+        if (!bundle) {
+            printf("[BPTestDumper] bundle failed to load!\n");
+            perror("bundle");
+            exit(1);
+        }
+        [bundle load];
+    }
+    NSLog(@"[BPTestDumper] Creating a default test suite with all suites of suites...");
+    XCTestSuite *defaultTestSuite = [XCTestSuite defaultTestSuite];
+    if (!defaultTestSuite) {
+        printf("[BPTestDumper] Failed to create a defaultTestSuite\n");
+        exit(1);
+    }
+    NSLog(@"[BPTestDumper] Successfully created a defaultTestSuite!!!");
+    NSMutableDictionary *testInfoDict = [[NSMutableDictionary alloc] init];
+    for (XCTestSuite *testSuite in [defaultTestSuite tests]) {
+        NSLog(@"[BPTestDumper] Parsing test bundle - %@; Number of test classes = %lu", [testSuite name], [testSuite testCaseCount]);
+        NSMutableDictionary *testClassesDict = [[NSMutableDictionary alloc] init];
+        for (XCTestSuite *testClass in [testSuite tests]) {
+            NSMutableArray *testCaseList = [[NSMutableArray alloc] init];
+            NSLog(@"[BPTestDumper] Parsing test class - %@; Number of test cases = %lu", [testClass name], [testClass testCaseCount]);
+            for (XCTest *testCase in [testClass tests]) {
+                NSLog(@"[BPTestDumper] Test case: %@", getSimpleName([testCase name]));
+                [testCaseList addObject:getSimpleName([testCase name])];
+            }
+            if ([testCaseList count] > 0) {
+                [testClassesDict setObject:(NSArray *)testCaseList forKey:[testClass name]];
+            }
+        }
+        if ([testClassesDict count] > 0) {
+            [testInfoDict setValue:testClassesDict forKey:[testSuite name]];
+        }
+    }
+    assert([NSJSONSerialization isValidJSONObject:testInfoDict]);
+    NSError *err;
+    NSData *json = [NSJSONSerialization dataWithJSONObject:testInfoDict
+                                                   options:NSJSONWritingPrettyPrinted
+                                                     error:&err];
+    if (!json) {
+        NSLog(@"[BPTestDumper] ERROR: %@", [err localizedDescription]);
+        exit(2);
+    }
+    if (![json writeToFile:outputFilePath atomically:YES]) {
+        NSLog(@"[BPTestDumper] Failed to dump the test list");
+    } else {
+        NSLog(@"[BPTestDumper] Successfully dumped a list of tests into the output file: %@", outputFilePath);
+        exit(2);
```

Comment:
I'm really confused by this code... above you check that the app's exit code is zero and you fail if it's not. Here you `exit(2)` on success and `0` on failure. How does this work?

Common Unix convention is to exit non-zero (typically `1`) on error and zero on success. We should keep that convention here.

Reply:
Oops, that's a typo. The `exit(2)` is supposed to be in the `if` clause.  I meant to exit with zero on success and non-zero on failure, exactly like what I added in a comment at the top.